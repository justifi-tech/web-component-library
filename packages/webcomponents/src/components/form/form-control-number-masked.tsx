import {
  Component,
  Host,
  h,
  Prop,
  Event,
  EventEmitter,
  Watch,
} from '@stencil/core';
import IMask, { InputMask } from 'imask';

@Component({
  tag: 'form-control-number-masked'
})
export class NumberInputMasked {
  textInput!: HTMLInputElement;

  @Prop() label: string;
  @Prop() name: any;
  @Prop() helpText?: string;
  @Prop() errorText?: string;
  @Prop() defaultValue: string;
  @Prop() inputHandler: (name: string, value: string) => void;
  @Prop() mask: string;
  @Prop() disabled: boolean;

  @Event() formControlInput: EventEmitter<any>;
  @Event() formControlBlur: EventEmitter<any>;

  private imask: InputMask<any> | null = null;

  updateInput(newValue: any) {
    if (this.imask && newValue) {
      this.imask.value = String(newValue);
    }
  }

  handleFormControlInput(event: any) {
    const target = event.target;
    const name = target.getAttribute('name');
    this.formControlInput.emit({ name: name, value: target.value });
  }

  private initializeIMask() {
    if (!this.textInput) return;

    this.imask = IMask(this.textInput, {
      mask: this.mask,
    });

    this.imask.on('accept', () => {
      const rawValue = this.imask.unmaskedValue;
      this.inputHandler(this.name, rawValue);
    });

    this.textInput.addEventListener('blur', () => this.formControlBlur.emit());
  }

  @Watch('defaultValue')
  handleDefaultValueChange(newValue: string) {
    this.updateInput(newValue);
  }

  componentDidLoad() {
    this.initializeIMask();
    this.updateInput(this.defaultValue);
  }

  disconnectedCallback() {
    this.imask?.destroy();
  }

  render() {
    return (
      <Host exportparts="label,input,input-invalid">
        <div class="form-group d-flex flex-column">
          <label part="label" class="form-label" htmlFor={this.name}>
            {this.label}
          </label>
          <input
            ref={el => (this.textInput = el as HTMLInputElement)}
            id={this.name}
            name={this.name}
            onInput={(event: any) => this.handleFormControlInput(event)}
            onBlur={() => this.formControlBlur.emit()}
            part={`input ${this.errorText && 'input-invalid'}`}
            class={this.errorText ? 'form-control is-invalid' : 'form-control'}
            type="text"
            disabled={this.disabled}
          />
          <form-control-help-text helpText={this.helpText} />
          <form-control-error-text errorText={this.errorText} />
        </div>
      </Host>
    );
  }
}