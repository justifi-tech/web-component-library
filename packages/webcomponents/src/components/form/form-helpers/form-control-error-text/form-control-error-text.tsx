import { Component, Prop, h } from '@stencil/core';

@Component({
  tag: 'form-control-error-text',
})
export class FormControlErrorText {
  @Prop() errorText: string;
  @Prop() name: string;

  render() {

    if (!this.errorText) {
      return null;
    }

    return (
      <small class='form-text text-danger' id={`form-error-text-${this.name}`}>
        {this.errorText}
      </small>
    );
  }
}
