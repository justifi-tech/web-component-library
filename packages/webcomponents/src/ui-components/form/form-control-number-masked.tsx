import {
  Component,
  Host,
  h,
  Prop,
  Event,
  EventEmitter,
  Watch,
  State,
} from '@stencil/core';
import IMask, { InputMask } from 'imask';
import { FormControlErrorText } from '../../ui-components';
import { input, inputDisabled, inputFocused, inputInvalid, label } from '../../styles/parts';

@Component({
  tag: 'form-control-number-masked'
})
export class NumberInputMasked {
  textInput!: HTMLInputElement;
  private imask: InputMask<any> | null = null;

  @State() isFocused: boolean = false;

  @Prop() label: string;
  @Prop() name: any;
  @Prop() helpText?: string;
  @Prop() errorText?: string;
  @Prop() defaultValue: string;
  @Prop() inputHandler: (name: string, value: string) => void;
  @Prop() mask: string;
  @Prop() disabled: boolean;

  @Watch('defaultValue')
  handleDefaultValueChange(newValue: string) {
    this.initializeIMask();
    this.updateInput(newValue);
  }

  @Event() formControlInput: EventEmitter<any>;
  @Event() formControlBlur: EventEmitter<any>;

  disconnectedCallback() {
    this.imask?.destroy();
  }

  componentDidLoad() {
    if (this.textInput) {
      this.initializeIMask();
      this.updateInput(this.defaultValue);
    }
  }

  private initializeIMask = () => {
    if (!this.textInput) return;

    this.imask = IMask(this.textInput, {
      mask: this.mask,
    });

    this.imask.on('accept', () => {
      const rawValue = this.imask?.unmaskedValue;
      if (rawValue !== undefined) {
        this.inputHandler(this.name, rawValue);
      }
    });

    this.textInput.addEventListener('blur', () => this.formControlBlur.emit());
  }

  handleFormControlInput = (event: any) => {
    const target = event.target;
    const name = target.getAttribute('name');
    const rawValue = this.imask?.unmaskedValue || target.value;
    this.formControlInput.emit({ name: name, value: rawValue });
  }

  updateInput = (newValue: any) => {
    if (this.imask && newValue) {
      this.imask.value = String(newValue);
    }
  }

  private get part() {
    if (this.errorText) {
      return inputInvalid;
    }
    if (this.disabled) {
      return inputDisabled;
    }
    if (this.isFocused) {
      return inputFocused;
    }
    return input;
  }

  render() {
    return (
      <Host>
        <div class="form-group d-flex flex-column">
          <div class="d-flex align-items-start gap-2">
            <label part={label} class="form-label" htmlFor={this.name}>
              {this.label}
            </label>
            <form-control-tooltip helpText={this.helpText} />
          </div>
          <input
            ref={el => (this.textInput = el as HTMLInputElement)}
            id={this.name}
            name={this.name}
            onFocus={() => (this.isFocused = true)}
            onBlur={() => {
              this.isFocused = false;
              this.formControlBlur.emit();
            }}
            onInput={this.handleFormControlInput}
            part={this.part}
            class={this.errorText ? 'form-control is-invalid' : 'form-control'}
            type="text"
            disabled={this.disabled}
          />
          <FormControlErrorText errorText={this.errorText} name={this.name} />
        </div>
      </Host>
    );
  }
}
