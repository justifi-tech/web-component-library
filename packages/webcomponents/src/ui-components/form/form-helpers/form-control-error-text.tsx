import { h } from '@stencil/core';

export const FormControlErrorText = (props: { errorText: string, name: string }) => {
  const { errorText, name } = props;

  if (!errorText) {
    return null;
  }

  return (
    <small class='form-text text-danger' id={`form-error-text-${name}`} part="input-error-text" >
      {errorText}
    </small>
  );
}
