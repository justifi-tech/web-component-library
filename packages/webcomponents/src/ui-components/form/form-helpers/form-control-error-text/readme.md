# form-control-error-text



<!-- Auto Generated Below -->


## Properties

| Property    | Attribute    | Description | Type     | Default     |
| ----------- | ------------ | ----------- | -------- | ----------- |
| `errorText` | `error-text` |             | `string` | `undefined` |
| `name`      | `name`       |             | `string` | `undefined` |


## Dependencies

### Used by

 - [form-control-checkbox](../..)
 - [form-control-date](../..)
 - [form-control-file](../..)
 - [form-control-file-v2](../..)
 - [form-control-monetary](../..)
 - [form-control-number](../..)
 - [form-control-number-masked](../..)
 - [form-control-select](../..)
 - [form-control-text](../..)
 - [form-control-textarea](../..)
 - [justifi-dispute-reason](../../../../components/dispute-management/dispute-response)

### Graph
```mermaid
graph TD;
  form-control-checkbox --> form-control-error-text
  form-control-date --> form-control-error-text
  form-control-file --> form-control-error-text
  form-control-file-v2 --> form-control-error-text
  form-control-monetary --> form-control-error-text
  form-control-number --> form-control-error-text
  form-control-number-masked --> form-control-error-text
  form-control-select --> form-control-error-text
  form-control-text --> form-control-error-text
  form-control-textarea --> form-control-error-text
  justifi-dispute-reason --> form-control-error-text
  style form-control-error-text fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
