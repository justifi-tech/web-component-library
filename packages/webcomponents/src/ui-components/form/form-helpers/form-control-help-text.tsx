import { h } from '@stencil/core';

export const FormControlHelpText = (props: { helpText: string, name: string }) => {
  const { helpText, name } = props;

  if (!helpText) {
    return null;
  }

  return (
    <small class='form-text text-muted' id={`form-help-text-${name}`} part="input-help-text">
      {helpText}
    </small>
  );
}
