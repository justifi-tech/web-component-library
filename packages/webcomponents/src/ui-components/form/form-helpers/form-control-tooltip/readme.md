# form-control-tooltip



<!-- Auto Generated Below -->


## Properties

| Property   | Attribute   | Description | Type     | Default     |
| ---------- | ----------- | ----------- | -------- | ----------- |
| `helpText` | `help-text` |             | `string` | `undefined` |


## Dependencies

### Used by

 - [form-control-date](../..)
 - [form-control-file](../..)
 - [form-control-file-v2](../..)
 - [form-control-number-masked](../..)
 - [form-control-select](../..)
 - [form-control-text](../..)
 - [justifi-additional-questions-form-step-core](../../../../components/business-forms/payment-provisioning/additional-questions)
 - [justifi-business-bank-account-form-step-core](../../../../components/business-forms/payment-provisioning/bank-account)
 - [justifi-business-owners-form-step-core](../../../../components/business-forms/payment-provisioning/business-owners)
 - [justifi-business-representative-form-inputs](../../../../components/business-forms/payment-provisioning/business-representative)
 - [justifi-legal-address-form-step-core](../../../../components/business-forms/payment-provisioning/legal-address-form)
 - [terminal-quantity-selector](../../../../components/order-terminals/terminal-quantity-selector)

### Depends on

- [custom-popper](../../../custom-popper)

### Graph
```mermaid
graph TD;
  form-control-tooltip --> custom-popper
  form-control-date --> form-control-tooltip
  form-control-file --> form-control-tooltip
  form-control-file-v2 --> form-control-tooltip
  form-control-number-masked --> form-control-tooltip
  form-control-select --> form-control-tooltip
  form-control-text --> form-control-tooltip
  justifi-additional-questions-form-step-core --> form-control-tooltip
  justifi-business-bank-account-form-step-core --> form-control-tooltip
  justifi-business-owners-form-step-core --> form-control-tooltip
  justifi-business-representative-form-inputs --> form-control-tooltip
  justifi-legal-address-form-step-core --> form-control-tooltip
  terminal-quantity-selector --> form-control-tooltip
  style form-control-tooltip fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
