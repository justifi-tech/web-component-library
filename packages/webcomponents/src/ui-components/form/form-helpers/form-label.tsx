import { h } from '@stencil/core';

export const FormLabel = ({ htmlFor, label, helpText }) => {
  return (
    <label
      part="label"
      class="form-label d-flex gap-2 align-items-center"
      htmlFor={htmlFor}
    >
      {label}
      <form-control-tooltip helpText={helpText} />
    </label>);
}
