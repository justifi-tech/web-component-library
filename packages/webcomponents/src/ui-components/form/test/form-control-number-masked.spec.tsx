import { h } from '@stencil/core';
import { newSpecPage } from "@stencil/core/testing";
import { NumberInputMasked } from "../form-control-number-masked";
import { FormControlErrorText } from '../form-helpers/form-control-error-text';
import { TooltipComponent } from '../form-helpers/form-control-tooltip/form-control-tooltip';

describe('form-control-number-masked', () => {
  const components = [NumberInputMasked, FormControlErrorText, TooltipComponent];
  const mockInputHandler = jest.fn();

  it('Renders with default props', async () => {
    const page = await newSpecPage({
      components: components,
      template: () => <form-control-number-masked label='Age' name='age' mask='00' />,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('Renders with all props provided', async () => {
    const page = await newSpecPage({
      components: components,
      template: () =>
        <form-control-number-masked
          label='Age'
          name='age'
          defaultValue='25'
          errorText='Invalid age'
          helpText='Enter your age'
          disabled
          mask='00'
          inputHandler={mockInputHandler}
        />
    });

    expect(page.root).toMatchSnapshot();
  });

  it('Updates the input value when defaultValue changes', async () => {
    // Initial render with the default value
    let defaultValue = '25';
    let page = await newSpecPage({
      components: components,
      template: () =>
        <form-control-number-masked
          label='Age'
          name='age'
          defaultValue={defaultValue}
          inputHandler={mockInputHandler}
          mask='00'
        />
    });
    let inputElement = page.root.querySelector('input');
    expect(inputElement.value).toBe('25');

    defaultValue = '30';

    page = await newSpecPage({
      components: components,
      template: () =>
        <form-control-number-masked
          label='Age'
          name='age'
          defaultValue={defaultValue}
          inputHandler={mockInputHandler}
          mask='00'
        />
    });

    inputElement = page.root.querySelector('input');
    expect(inputElement.value).toBe('30');
  });

  it('Handles user input correctly', async () => {
    const page = await newSpecPage({
      components: components,
      template: () =>
        <form-control-number-masked
          label='Age'
          name='age'
          defaultValue='25'
          errorText='Invalid age'
          helpText='Enter your age'
          disabled
          mask='00'
          inputHandler={mockInputHandler}
        />
    });

    const inputElement = page.root.querySelector('input');
    const testValue = '30';

    inputElement.value = testValue;
    await inputElement.dispatchEvent(new Event('input'));

    expect(inputElement.value).toBe(testValue);
  });

  it('Emits formControlInput event on input', async () => {
    const page = await newSpecPage({
      components: components,
      template: () =>
        <form-control-number-masked
          label='Age'
          name='age'
          defaultValue='25'
          inputHandler={mockInputHandler}
          mask='00'
        />
    });

    const inputEventSpy = jest.fn();
    page.root.addEventListener('formControlInput', inputEventSpy);

    const inputElement = page.root.querySelector('input');
    inputElement.value = '25';

    inputElement.dispatchEvent(new Event('input', { bubbles: true, composed: true }));

    expect(inputEventSpy).toHaveBeenCalledWith(expect.objectContaining({
      detail: {
        name: 'age',
        value: '25',
      }
    }));
  });

  it('Emits formControlBlur event on blur', async () => {
    const page = await newSpecPage({
      components: components,
      template: () =>
        <form-control-number-masked
          label='Age'
          name='age'
          inputHandler={mockInputHandler}
          mask='00'
        />
    });
    const blurEventSpy = jest.fn();
    page.win.addEventListener('formControlBlur', blurEventSpy);

    const inputElement = page.root.querySelector('input');
    await inputElement.dispatchEvent(new CustomEvent('blur'));

    expect(blurEventSpy).toHaveBeenCalled();
  });

  it('Disables input when disabled prop is provided', async () => {
    const page = await newSpecPage({
      components: components,
      template: () =>
        <form-control-number-masked
          label='Age'
          name='age'
          disabled
          inputHandler={mockInputHandler}
          mask='00'
        />
    });

    const inputElement = page.root.querySelector('input');
    expect(inputElement.disabled).toBe(true);
  });

  it('Shows help text when helpText prop is provided', async () => {
    const page = await newSpecPage({
      components: components,
      template: () =>
        <form-control-number-masked
          label='Age'
          name='age'
          helpText='Enter your age'
          inputHandler={mockInputHandler}
          mask='00'
        />
    });

    const tooltipComponent = page.root.querySelector('form-control-tooltip');
    expect(tooltipComponent).not.toBeNull();

    const tooltipIcon = tooltipComponent.querySelector('.bi-question-square');
    expect(tooltipIcon).not.toBeNull();

    const tooltipElement = tooltipComponent.querySelector('.tooltip');
    expect(tooltipElement).not.toBeNull();

    const tooltipText = tooltipElement.querySelector('.tooltip-inner');
    expect(tooltipText.textContent).toBe('Enter your age');
  });

  it('Shows error and applies error styling when error prop is provided', async () => {
    const page = await newSpecPage({
      components: components,
      template: () =>
        <form-control-number-masked
          label='Age'
          name='age'
          errorText='Invalid age'
          inputHandler={mockInputHandler}
          mask='00'
        />
    });

    const errorTextComponent = page.root.querySelector('#form-error-text-age');
    expect(errorTextComponent).not.toBeNull();

    expect(errorTextComponent.textContent).toBe('Invalid age');

    const inputElement = page.root.querySelector('input');
    expect(inputElement.classList.contains('is-invalid')).toBe(true);
  });

  it('Emits unmasked value for SSN mask input', async () => {
    const mockHandler = jest.fn();
    const page = await newSpecPage({
      components: components,
      template: () =>
        <form-control-number-masked
          label='SSN'
          name='identification_number'
          inputHandler={mockHandler}
          mask='000-00-0000'
        />
    });

    const inputEventSpy = jest.fn();
    page.root.addEventListener('formControlInput', inputEventSpy);

    const inputElement = page.root.querySelector('input');
    
    // Simulate entering SSN with dashes
    inputElement.value = '123-45-6789';
    await inputElement.dispatchEvent(new Event('input', { bubbles: true, composed: true }));

    // The formControlInput event should emit the unmasked value (digits only)
    // Note: In the test environment, IMask might not be fully functional,
    // but the logic should attempt to use unmaskedValue when available
    expect(inputEventSpy).toHaveBeenCalled();
    
    // The event should have been called with either the unmasked value or fall back to input value
    const lastCall = inputEventSpy.mock.calls[inputEventSpy.mock.calls.length - 1][0];
    expect(lastCall.detail.name).toBe('identification_number');
    // In real usage, this would be '123456789' (unmasked), but in test it might fall back to input value
    expect(lastCall.detail.value).toBeDefined();
  });
});
