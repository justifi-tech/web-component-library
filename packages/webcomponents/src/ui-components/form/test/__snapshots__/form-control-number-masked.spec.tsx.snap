// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`form-control-number-masked Renders with all props provided 1`] = `
<form-control-number-masked exportparts="label,input,input-invalid">
  <div class="d-flex flex-column form-group">
    <div class="align-items-start d-flex gap-2">
      <label class="form-label" htmlfor="age" part="label">
        Age
      </label>
      <form-control-tooltip>
        <div class="tooltip-container">
          <svg class="bi bi-question-square tooltip-icon" fill="currentColor" height="16" part="tooltip-icon" viewBox="0 0 16 16" width="16" xmlns="http://www.w3.org/2000/svg">
            <path d="M14 1a1 1 0 0 1 1 1v12a1 1 0 0 1-1 1H2a1 1 0 0 1-1-1V2a1 1 0 0 1 1-1zM2 0a2 2 0 0 0-2 2v12a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V2a2 2 0 0 0-2-2z"></path>
            <path d="M5.255 5.786a.237.237 0 0 0 .241.247h.825c.138 0 .248-.113.266-.25.09-.656.54-1.134 1.342-1.134.686 0 1.314.343 1.314 1.168 0 .635-.374.927-.965 1.371-.673.489-1.206 1.06-1.168 1.987l.003.217a.25.25 0 0 0 .25.246h.811a.25.25 0 0 0 .25-.25v-.105c0-.718.273-.927 1.01-1.486.609-.463 1.244-.977 1.244-2.056 0-1.511-1.276-2.241-2.673-2.241-1.267 0-2.655.59-2.75 2.286m1.557 5.763c0 .533.425.927 1.01.927.609 0 1.028-.394 1.028-.927 0-.552-.42-.94-1.029-.94-.584 0-1.009.388-1.009.94"></path>
          </svg>
          <custom-popper placement="top" trigger="hover">
            <div class="bs-tooltip-top tooltip" part="tooltip" role="tooltip">
              <div class="tooltip-inner" part="tooltip-inner">
                Enter your age
              </div>
            </div>
          </custom-popper>
        </div>
      </form-control-tooltip>
    </div>
    <input class="form-control is-invalid" disabled="" id="age" name="age" part="input input-invalid" type="text" value="25">
    <small class="form-text text-danger" id="form-error-text-age" part="input-error-text">
      Invalid age
    </small>
  </div>
</form-control-number-masked>
`;

exports[`form-control-number-masked Renders with default props 1`] = `
<form-control-number-masked exportparts="label,input,input-invalid">
  <div class="d-flex flex-column form-group">
    <div class="align-items-start d-flex gap-2">
      <label class="form-label" htmlfor="age" part="label">
        Age
      </label>
      <form-control-tooltip></form-control-tooltip>
    </div>
    <input class="form-control" id="age" name="age" part="input undefined" type="text">
  </div>
</form-control-number-masked>
`;
